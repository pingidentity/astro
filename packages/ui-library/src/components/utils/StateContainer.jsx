import React from "react";
import PropTypes from "prop-types";
import Utils from "../../util/Utils";

const mapNothing = () => ({});

const defaultGetPassedProps = props => props;
const containerGetPassedProps = props => props.passedProps;

export const createProgressiveState = stateDefs => ({
    passedProps,
    initialState,
    boundSetState,
    getPassedProps = defaultGetPassedProps,
}) => {
    const isControlled = name => passedProps[name] === undefined;

    // figure out which props will be controlled and create initial state for those
    const pState = stateDefs.reduce((state, { name, initial }) => {
        if (isControlled(name)) {
            return {
                ...state,
                [name]: initialState[name] === undefined ? initial : initialState[name],
                // the initialState prop should override the defined initial value
                // but only if this is a controlled prop
            };
        }
        return state;
    }, {});

    // make sure the keys provided in initialState are actually being controlled
    if (!Utils.isProduction()) {
        Object.keys(initialState).forEach(key => {
            if (pState[key] === undefined) {
                console.warn(
                    `Initial state provided for prop "${key}" that is not being controlled by the component`
                );
            }
        });
    }

    // generate custom callbacks with transform functions
    const makeTransformCallbacks = (name, callbacks = []) =>
        callbacks.reduce(
            (transformCallbacks, { name: cbName, transform, passTransformedValue = false, }) => ({
                ...transformCallbacks,
                [cbName]: (value, e) => {
                    // set the state with the transformed value
                    boundSetState((state, props) => {
                        const transformedValue = transform(value, state[name], getPassedProps(props));

                        if (passedProps[cbName]) {
                            passedProps[cbName](passTransformedValue ? transformedValue : value, e);
                        }

                        return { [name]: transformedValue };
                    });
                }
            }), {}
        );

    // get the callbacks for a single state definition
    const makeCallbacksFromDef = ({ name, setter, callbacks }) => {
        // add any custom callbacks
        const customTransformCallbacks = makeTransformCallbacks(name, callbacks);

        // return with setter if defined
        return setter ? {
            [setter]: (value, e) => {
                boundSetState({ [name]: value });
                // if a callback is provided, we'll still execute it
                if (passedProps[setter]) {
                    passedProps[setter](value, e);
                }
            },
            ...customTransformCallbacks
        } : customTransformCallbacks;
    };

    // create all the callbacks
    const pCallbacks = stateDefs.reduce((returnCallbacks, def) => {
        if (!isControlled(def.name)) {
            return returnCallbacks;
        }
        return { ...returnCallbacks, ...makeCallbacksFromDef(def) };
    }, {});

    return [pState, pCallbacks];
};

class StateContainer extends React.Component {
    static propTypes = {
        stateDefs: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired, // name of the props
                initial: PropTypes.any.isRequired, // starting value
                setter: PropTypes.string, // name of the callback in the target component
                callbacks: PropTypes.arrayOf( // when you need something different than a simple setter
                    PropTypes.shape({
                        name: PropTypes.string,
                        // Decides whether to pass the transformed value to the callback or just the value that
                        // the base component passes.
                        passTransformedValue: PropTypes.bool,
                        transform: PropTypes.func, // this function will be applied to the value and current state
                    })
                ),
            })
        ),
        initialState: PropTypes.object,
        passedProps: PropTypes.object,
        mapToProps: PropTypes.func,
    };

    static defaultProps = {
        stateDefs: [],
        initialState: {},
        passedProps: {},
        mapToProps: mapNothing,
    };

    constructor({ stateDefs, initialState, passedProps }) {
        // the constructor has two main goals
        // 1) populate the state for any props that will be controlled
        // 2) create callback functions for those props
        super();

        const [ state, callbacks ] = createProgressiveState(stateDefs)({
            initialState, passedProps, boundSetState: this.setState.bind(this), getPassedProps: containerGetPassedProps
        });

        this.state = state;
        this.callbacks = callbacks;
    }

    render() {
        const { children, mapToProps, passedProps } = this.props;

        // alert dev if there are state/prop conflicts
        if (!Utils.isProduction()) {
            Object.keys(this.state).forEach(key => {
                if (passedProps[key] !== undefined) {
                    console.warn(
                        `The prop "${key}" is being controlled by the component because it was not initially set. ` +
                        `The newly-received value will be ignored.`
                    );
                }
            });
        }

        return children({ ...passedProps, ...this.callbacks, ...this.state, ...mapToProps(this.state, passedProps) });
    }
}

export const inStateContainer = (stateDefs, mapToProps) => WrappedComponent => ({ initialState, ...props }) => (
    <StateContainer stateDefs={stateDefs} mapToProps={mapToProps} initialState={initialState} passedProps={props}>
        {containerProps => <WrappedComponent {...containerProps} />}
    </StateContainer>
);

export const toggleTransform = (value, current) => (!current);

export default StateContainer;