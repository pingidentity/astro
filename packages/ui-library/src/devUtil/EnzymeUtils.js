import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
