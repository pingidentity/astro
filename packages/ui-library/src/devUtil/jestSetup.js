
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
