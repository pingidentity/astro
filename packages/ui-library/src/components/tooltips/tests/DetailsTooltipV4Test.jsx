window.__DEV__ = true;

jest.mock("popper.js");
jest.mock("react-portal");

import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import StateContainer from "../../utils/StateContainer";

describe("DetailsTooltip", function () {

    const React = require("react"),
        ReactDOM = require("react-dom"),
        ReactTestUtils = require("react-dom/test-utils"),
        TestUtils = require("../../../testutil/TestUtils"),
        DetailsTooltip = require("../DetailsTooltip"),
        Wrapper = TestUtils.UpdatePropsWrapper,
        _ = require("underscore");

    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    function getComponent (opts, alwaysOpen = true) {
        opts = _.defaults(opts, {
            title: "Title",
            label: "Label",
            open: alwaysOpen ? true : undefined,
            onValueChange: jest.fn(),
            onToggle: jest.fn(),
        });

        return ReactTestUtils.renderIntoDocument(
            <DetailsTooltip {...opts}><p>what ever callout content is</p></DetailsTooltip>);
    }
    beforeEach(function () {
        window.addEventListener.mockClear();
        window.removeEventListener.mockClear();
        beforeEach(function() {
            global.getSelection = jest.fn();
            global.getSelection.mockReturnValue({
                toString: () => "",
            });
        });
    });
    afterEach(function () {
        delete process.env.NODE_ENV;
    });

    it("data-id's don't change", () => {
        const labels = [
            { value: jest.fn(), label: "Save" },
            { value: jest.fn(), label: "Save More" },
        ];
        mountSnapshotDataIds(
            <DetailsTooltip
                cancelLabel="UH OH"
                label="OH JEEZ"
                primaryLabels={labels}
                secondaryLabels={labels}
                showClose
                title="OH BISCUITS"
            />
        );
    });

    it("managed component that starts open", function () {
        const component = getComponent({ initialState: { open: true } }, false);
        expect(ReactTestUtils.findRenderedComponentWithType(component, StateContainer).state.open).toBe(true);
    });

    it("managed component", function () {
        const component = getComponent({ initialState: { open: false } }, false);
        const link = TestUtils.findRenderedDOMNodeWithDataId(component, "action-btn");
        const manager = ReactTestUtils.findRenderedComponentWithType(component, StateContainer);
        expect(manager.state.open).toBe(false);

        ReactTestUtils.Simulate.click(link, {});
        expect(manager.state.open).toBe(true);
    });

    it("is rendering open state", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const title = TestUtils.findRenderedDOMNodeWithDataId(details, "details-title");
        const content = TestUtils.findRenderedDOMNodeWithDataId(details, "details-content").children[0];
        const body = TestUtils.findRenderedDOMNodeWithDataId(details, "details-body");
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "action-btn");

        //DOM to contain actual content
        expect(content.children.length).toBe(3);
        expect(body.children[0].nodeName).toBe("P");
        expect(body.children[0].textContent).toBe("what ever callout content is");

        //action link
        expect(link.textContent).toBe("Action");

        //title
        expect(title.textContent).toBe("Title");
    });

    it("doesnt render title container if none exists", function () {
        const component = getComponent({ title: "" });
        const title = TestUtils.findRenderedDOMNodeWithDataId(component, "details-title");

        expect(title).toBe(null);
    });

    it("is rendering open state with custom styles and no close control", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true} titleClassName="title"
                contentClassName="content" showClose={false} labelClassName="label" className="extra">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const title = TestUtils.findRenderedDOMNodeWithDataId(details, "details-title");
        const content = TestUtils.findRenderedDOMNodeWithDataId(details, "details-content").children[0];
        const body = TestUtils.findRenderedDOMNodeWithDataId(details, "details-body");
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "action-btn");
        const closeLink = TestUtils.scryRenderedDOMNodesWithDataId(details, "details-close");

        //make sure root div got extra CSS classes
        expect(ReactDOM.findDOMNode(component).getAttribute("class")).toContain("extra");

        //DOM to contain actual content
        expect(content.children.length).toBe(2);
        expect(body.children[0].nodeName).toBe("P");
        expect(body.children[0].textContent).toBe("what ever callout content is");

        //no close link
        expect(closeLink.length).toBe(0);

        //action link
        expect(link.textContent).toBe("Action");

        //title
        expect(title.textContent).toBe("Title");
    });

    it("is rendering closed state", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={false}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const content = TestUtils.scryRenderedDOMNodesWithDataId(details, "details-content");

        //no DOM content
        expect(content.length).toBe(0);
    });

    it("is notifying on toggle", function () {

        const callback = jest.fn();

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={false} onToggle={callback}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "action-btn");

        ReactTestUtils.Simulate.click(link, {});

        expect(callback).toBeCalled(); //make sure callback was triggered
    });

    it("is notifying on toggle via content click", function () {

        const callback = jest.fn();

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true}
                onToggle={callback} showClose={false} hideOnClick={true}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const content = TestUtils.findRenderedDOMNodeWithDataId(details, "details-content");

        ReactTestUtils.Simulate.click(content, {});

        expect(callback).toBeCalled(); //make sure callback was triggered
    });

    it("is notifying on toggle via close control", function () {

        const callback = jest.fn();

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true} onToggle={callback}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "details-close");

        ReactTestUtils.Simulate.click(link, {});

        expect(callback).toBeCalled(); //make sure callback was triggered
    });

    // Remove this once V4 is released. Calling component methods is a bad look.
    it("when stateless, close method doesn't fire an error", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        component.close();
    });

    it("is supports disabling", function () {
        const callback = jest.fn();

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true}
                title="Title" label="Action" open={false} onToggle={callback} disabled={true}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "action-btn");

        //make disabled css styling was added
        expect(link.getAttribute("class")).toContain("disabled");

        ReactTestUtils.Simulate.click(link, {});

        expect(callback).not.toBeCalled(); //make sure callback was not triggered
    });

    it("registers global listener on mount if component is open", function () {
        //let's override defer or execute func immediately for tests
        _.defer = function (func) {
            func();
        };

        ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        expect(TestUtils.mockCallsContains(window.addEventListener, "click")).toBe(true);
        expect(TestUtils.mockCallsContains(window.addEventListener, "keydown")).toBe(true);
    });

    it("unregister global listeners on unmount", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        //trigger unmount
        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(component).parentNode);

        expect(TestUtils.mockCallsContains(window.removeEventListener, "click")).toBe(true);
        expect(TestUtils.mockCallsContains(window.removeEventListener, "keydown")).toBe(true);
    });

    //TODO no idea why this isn't working. can console log the full stack, it just isn't being called
    it("triggers callback when clicked outside", function () {
        const callback = jest.fn();

        ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true} onToggle={callback}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "click")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        //click outside
        handler(e);

        expect(callback).toBeCalled();
    });

    it("triggers callback when escape is pressed", function () {
        const callback = jest.fn();

        ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true} onToggle={callback}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "keydown")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn(),
            keyCode: 27,
        };

        //press escape
        handler(e);

        expect(callback).toBeCalled();
    });

    it("doesn't trigger callback when a key other than escape is pressed", function () {
        const callback = jest.fn();

        ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={true} onToggle={callback}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "keydown")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn(),
            keyCode: 10,
        };

        //press escape
        handler(e);

        expect(callback).not.toBeCalled();
    });

    it("doesn't error when handling outside click but not open", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={false}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        ReactTestUtils.findRenderedComponentWithType(
            component,
            DetailsTooltip.DetailsTooltipStateless
        )._handleGlobalClick();
    });

    it("doesn't error when handling escape click but not open", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip stateless={true} title="Title" label="Action" open={false}>
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        ReactTestUtils.findRenderedComponentWithType(
            component,
            DetailsTooltip.DetailsTooltipStateless
        )._handleGlobalKeyDown({ keyCode: 27 });
    });

    it("unregister listener when transitioning from open to closed", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <Wrapper type={DetailsTooltip}
                stateless={true} title="Title" label="Action" open={true} onToggle={jest.genMockFn}>
                <p>what ever callout content is</p>
            </Wrapper>);

        component._setProps({ open: false });

        expect(TestUtils.mockCallsContains(window.removeEventListener, "click")).toBe(true);
        expect(TestUtils.mockCallsContains(window.removeEventListener, "keydown")).toBe(true);
    });

    it("render component with data-id", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip data-id="detailsTooltipWithDataId" title="Title" label="Action">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const element = TestUtils.findRenderedDOMNodeWithDataId(component, "detailsTooltipWithDataId");

        expect(element).toBeDefined();
    });

    it("render component with default data-id", function () {
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const element = TestUtils.findRenderedDOMNodeWithDataId(component, "details-tooltip");

        expect(element).toBeDefined();
    });

    it("is rendering label styling with labelClassName", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true} labelClassName="customLabel">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const link = TestUtils.findRenderedDOMNodeWithDataId(details, "action-btn");

        //make sure anchor tag got customLabel CSS classes
        expect(ReactDOM.findDOMNode(link).getAttribute("class")).toContain("customLabel");
    });

    it("is rendering content styling with contentClassName", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true} contentClassName="customContent">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const content = TestUtils.findRenderedDOMNodeWithDataId(details, "details-content");

        //make sure context div got customContent CSS classes
        expect(ReactDOM.findDOMNode(content).getAttribute("class")).toContain("customContent");
    });

    it("is rendering title styling with titleClassName", function () {

        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip title="Title" label="Action" open={true} titleClassName="customTitle">
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const details = TestUtils.findRenderedComponentWithType(component, DetailsTooltip);
        const title = TestUtils.findRenderedDOMNodeWithDataId(details, "details-title");

        //make sure context div got customContent CSS classes
        expect(ReactDOM.findDOMNode(title).getAttribute("class")).toContain("customTitle");
    });

    it("Renders buttons when optional buttons are provided", function () {
        const secondaryArr = [
                { value: jest.fn(), label: "One" },
                { value: jest.fn(), label: "Two" }
            ],
            primaryArr = [
                { value: jest.fn(), label: "Save" },
                { value: jest.fn(), label: "Save More" },
            ];
        const component = ReactTestUtils.renderIntoDocument(
            <DetailsTooltip
                title="Buttons provided"
                label="Action"
                open={true}
                secondaryLabels={secondaryArr}
                primaryLabels={primaryArr}
                cancelLabel="Cancel">
                <p>What ever callout content is</p>
            </DetailsTooltip>
        );
        const saveBtn = TestUtils.scryRenderedDOMNodesWithClass(component, "primary"),
            cancelBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "cancel-action"),
            secondaryBtns = TestUtils.scryRenderedDOMNodesWithTag(component, "button");

        expect(secondaryBtns).toBeTruthy();
        expect(saveBtn).toBeTruthy();
        expect(cancelBtn).toBeTruthy();

        // check length of buttons
        expect(saveBtn.length).toBe(2);
        expect(TestUtils.scryRenderedDOMNodesWithClass(component, "cancel").length).toBe(3);
        expect(secondaryBtns.length).toBe(4);

        // check button text
        expect(secondaryBtns[0].textContent).toBe("One");
        expect(secondaryBtns[1].textContent).toBe("Two");
        expect(cancelBtn.textContent).toBe("Cancel");
        expect(secondaryBtns[2].textContent).toBe("Save");
        expect(secondaryBtns[3].textContent).toBe("Save More");
    });

    it("is rendering with the right x-placement attribute based on the placement prop", function() {
        const positionedComponent = placement => ReactTestUtils.renderIntoDocument(
            <DetailsTooltip
                stateless={true}
                title="Title"
                label="Action"
                open={true}
                placement={placement}
            >
                <p>what ever callout content is</p>
            </DetailsTooltip>
        );

        const positioningProp = position => ReactTestUtils.findRenderedComponentWithType(
            positionedComponent(position),
            DetailsTooltip.DetailsTooltipStateless
        ).popperContainer.props.placement;

        expect(positioningProp("top right")).toBe("top-start");
        expect(positioningProp("bottom right")).toBe("bottom-start");
        expect(positioningProp("top left")).toBe("top-end");
        expect(positioningProp("bottom left")).toBe("bottom-end");
        expect(positioningProp("top")).toBe("top");
        expect(positioningProp("bottom")).toBe("bottom");
    });

});
