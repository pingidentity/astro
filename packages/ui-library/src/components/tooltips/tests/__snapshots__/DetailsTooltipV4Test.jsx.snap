// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsTooltip data-id's don't change 1`] = `
Array [
  "details-tooltip",
  "action-btn",
]
`;
